pkg/distributor/rate_strategy.go

```diff
@@ -0,0 +1,124 @@
+package distributor
+
+import (
+	"math"
+
+	"golang.org/x/time/rate"
+
+	"github.com/cortexproject/cortex/pkg/util/limiter"
+	"github.com/cortexproject/cortex/pkg/util/validation"
+)
+
+// ReadLifecycler represents the read interface to the lifecycler.
+type ReadLifecycler interface {
+	HealthyInstancesCount() int
+}
+
+type localStrategy struct {
+	baseStrategy limiter.RateLimiterStrategy
+}
+
+func newLocalRateStrategy(baseStrategy limiter.RateLimiterStrategy) limiter.RateLimiterStrategy {
+	return &localStrategy{
+		baseStrategy: baseStrategy,
+	}
+}
+
+func (s *localStrategy) Limit(tenantID string) float64 {
+	return s.baseStrategy.Limit(tenantID)
+}
+
+func (s *localStrategy) Burst(tenantID string) int {
+	// The meaning of burst doesn't change for the global strategy, in order
+	// to keep it easier to understand for users / operators.
+	return s.baseStrategy.Burst(tenantID)
+}
+
+type globalStrategy struct {
+	baseStrategy limiter.RateLimiterStrategy
+	ring         ReadLifecycler
+}
+
+func newGlobalRateStrategy(baseStrategy limiter.RateLimiterStrategy, ring ReadLifecycler) limiter.RateLimiterStrategy {
+	return &globalStrategy{
+		baseStrategy: baseStrategy,
+		ring:         ring,
+	}
+}
+
+func (s *globalStrategy) Limit(tenantID string) float64 {
+	numDistributors := s.ring.HealthyInstancesCount()
+	limit := s.baseStrategy.Limit(tenantID)
+
+	if numDistributors == 0 || limit == float64(rate.Inf) {
+		return limit
+	}
+	return limit / float64(numDistributors)
```

Comment:
this is very similar to what we had before but I have my doubts. In any case, your PR attempts to add an experimental feature. I think you should not touch ingestion_rate_strategy.go, you should add a new request_rate_strategy.go. Leave the refactoring for later.

Reply:
In fact, if it is really completely independent, do not modify the current ingestion_rate_strategy.go code. The new code will be very ugly, such as having to be named globalRequestStrategy/localRequestStrategy, but in fact there is a public globalRateStrategy/localRateStrategy that can be reused. Even if it is an experimental feature or change, the existence of similar code is unbearable (and the follow-up, when, and who will update it is unknown).

https://github.com/cortexproject/cortex/blob/78e6da8ccb9aa085a4045fd96b1ef01aff4ebabd/pkg/distributor/ingestion_rate_strategy.go#L33

No offense but frankly, we should focus on Cortex's faster iterations and stronger capabilities, especially in comparison to Mimir. :)

Reply:
In the case of ensuring compatibility, I think we should adopt a slightly aggressive strategy to operate Cortex, and fewer people will maintain and pay attention to it, which will only make it perish.

And I am very much looking forward to and willing to promote Cortex together. I think this is a project with great potential.

Reply:
I sincerely and most certainly appreciate your contributions.  Cortex definitely needs more contributions for sure.
I will take another look